regression-tests.dnsdist/dnsdisttests.py

```diff
@@ -142,42 +142,54 @@ def _ResponderIncrementCounter(cls):
             cls._responsesCounter[threading.currentThread().name] = 1
 
     @classmethod
-    def _getResponse(cls, request, fromQueue, toQueue):
+    def _getResponse(cls, request, fromQueue, toQueue, synthesize=None):
         response = None
         if len(request.question) != 1:
             print("Skipping query with question count %d" % (len(request.question)))
             return None
         healthCheck = str(request.question[0].name).endswith(cls._healthCheckName)
         if healthCheck:
             cls._healthCheckCounter += 1
+            response = dns.message.make_response(request)
         else:
             cls._ResponderIncrementCounter()
             if not fromQueue.empty():
-                response = fromQueue.get(True, cls._queueTimeout)
-                if response:
-                    response = copy.copy(response)
-                    response.id = request.id
-                    toQueue.put(request, True, cls._queueTimeout)
+                toQueue.put(request, True, cls._queueTimeout)
+                if synthesize is None:
+                    response = fromQueue.get(True, cls._queueTimeout)
+                    if response:
+                        response = copy.copy(response)
+                        response.id = request.id
 
         if not response:
-            if healthCheck:
+            if synthesize is not None:
                 response = dns.message.make_response(request)
+                response.set_rcode(synthesize)
             elif cls._answerUnexpected:
                 response = dns.message.make_response(request)
                 response.set_rcode(dns.rcode.SERVFAIL)
 
         return response
 
     @classmethod
-    def UDPResponder(cls, port, fromQueue, toQueue, ignoreTrailing=False):
+    def UDPResponder(cls, port, fromQueue, toQueue, trailingDataResponse=False):
+        ignoreTrailing = trailingDataResponse is True
         sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
         sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
         sock.bind(("127.0.0.1", port))
         while True:
             data, addr = sock.recvfrom(4096)
-            request = dns.message.from_wire(data, ignore_trailing=ignoreTrailing)
-            response = cls._getResponse(request, fromQueue, toQueue)
-
+            forceRcode = None
+            try:
+                request = dns.message.from_wire(data, ignore_trailing=ignoreTrailing)
+            except dns.message.TrailingJunk as e:
+                if trailingDataResponse is False:
+                    raise
+                print("UDP query with trailing data, synthesizing response")
```

Comment:
The existing structure was already a bit confusing, and I was trying to preserve it as much as possible (which unfortunately didn't help). But the idea is to extend the existing `ignoreTrailing` parameter into `trailingDataResponse` as follows:
* `False`: reject queries with trailing data (_preexisting behavior_)
* `True`: ignore trailing data after queries (_preexisting behavior_)
* anything else: interpret as RCODE to be sent in response to queries with trailing data (_new behavior, used by `TestTrailingDataToBackend` in regression-tests.dnsdist/test_Trailing.py_)
We will never pass `synthesize=True` to `_getResponse`, because `forceRcode` is only set to `trailingDataResponse` when dns.message.from_wire raises dns.message.TrailingJunk, which cannot happen when it is ignoring trailing data.

I have added comments and safety-enforcing code.